### PR TITLE
Fix reviewer cost and quota diagnostics

### LINKED ISSUE #84: Track reviewer provider cost, quota, and explicit purchase/upgrade workflow
## Context

Follow-up split from #77 / PR #82.

Reviewer UX failures can be caused by auth, provider outages, prompt/request size, local tunnel readiness, quota exhaustion, usage tier limits, or actual spend limits. PR #82 improves runtime diagnostics, but it intentionally does not let agents make financial transactions or mutate billing state.

The desired future workflow is to know when cost/quota is the limiting factor and, if appropriate, support an explicit user-authorized path for increasing a usage tier or purchasing credits.

## Scope

- Diagnose whether reviewer failures are cost/quota/usage-tier related versus auth, timeout, request-size, or provider readiness failures.
- Normalize cost/quota fields across Claude, Gemini, Kimi, Grok Web, DeepSeek, and GLM where providers expose safe metadata.
- Persist privacy-safe cost/usage telemetry in JobRecords without storing secrets, invoices, cookies, raw provider payloads, or payment details.
- Design an explicit transaction workflow for billing actions:
  - no silent purchases,
  - no automatic tier upgrades,
  - no stored payment details,
  - require a clear user approval prompt with provider, account, amount/tier, and expected effect.
- Clearly distinguish primary CODEX_HOME / secondary CODEX_HOME diagnostics from provider billing accounts.

## Acceptance criteria

- Reviewer JobRecords can distinguish quota/usage-tier/cost-limit failures from auth, timeout, request-size, and provider-readiness failures when the provider exposes enough information.
- Doctor/setup commands report cost/quota readiness only from privacy-safe metadata and never leak credentials or billing artifacts.
- Any purchase/upgrade command, if implemented, is opt-in and requires explicit user confirmation at execution time.
- Tests cover at least one quota/cost classification path per applicable provider family.
- Documentation states that financial transactions are out of automatic review flow and require user approval.

## Links

- Parent runtime diagnosability work: #77
- Implementation PR for current diagnostics: #82
- Oversized scope/provider-size follow-up: #83

### PR DESCRIPTION
**Pull request by seungpyoson**
## Summary

Closes #84.

- classify reviewer quota, billing, credit-limit, and usage-tier failures as `usage_limited` across Claude, Gemini, Kimi, Grok, and direct API reviewer paths
- keep transient direct API HTTP 429 responses as `rate_limited`, while quota-bearing 429 responses remain `usage_limited`
- keep HTTP 501 compatibility/configuration failures as `provider_error`, not transient provider unavailability
- share usage-limit detection through synchronized `usage-limit.mjs` copies, with tests preventing drift across provider paths
- exclude those intentionally synchronized packaging copies from Sonar CPD, with a CI contract test covering the exclusion list
- persist only bounded runtime diagnostics for cost/quota failure cases, with explicit `billing_mutation` values and no raw prompts/source/payment artifacts in `runtime_diagnostics.cost_quota`
- preserve the PR #89 safe-Git diagnostics after rebase: `git_binary_rejected -> git_binary_policy` remains present alongside `usage_limited -> cost_quota_usage_limit` in API Reviewers and Grok

## Current Refs

- Base: `c6cebc4823d0a71b4798b07f7063215868f28709`
- Head: `635240497ba1a19b5245cbf07dbe98dcdb481a07`

## Verification

Local verification on current head:

- `node --test tests/smoke/api-reviewers.smoke.test.mjs tests/smoke/grok-web.smoke.test.mjs` -> 166 passed, 0 failed
- `node --test tests/unit/coverage-script.test.mjs tests/unit/usage-limit.test.mjs tests/unit/plugin-copies-in-sync.test.mjs` -> 50 passed, 0 failed
- `node --test tests/unit/usage-limit.test.mjs tests/unit/claude-dispatcher.test.mjs tests/unit/gemini-dispatcher.test.mjs tests/unit/kimi-dispatcher.test.mjs tests/unit/plugin-copies-in-sync.test.mjs tests/unit/lib-imports.test.mjs tests/unit/docs-contracts.test.mjs` -> 274 passed, 0 failed
- `node --test tests/unit/ci-workflow.test.mjs tests/unit/plugin-copies-in-sync.test.mjs tests/unit/usage-limit.test.mjs` -> 52 passed, 0 failed
- `npm run lint:sync` -> passed
- `git diff --check` -> passed
- `npm test` -> 1126 passed, 6 skipped, 0 failed
- pre-commit `npm test` -> 1126 passed, 6 skipped, 0 failed

GitHub checks pass on current head `6352404`: SonarCloud, lint, test, smoke api-reviewers, smoke claude, smoke gemini, smoke grok, and smoke kimi.

## Review Status

Earlier external review results in this PR body and comment history targeted older heads and are stale after the rebase onto merged PR #89 and subsequent Gemini Code Assist fixes.

Gemini Code Assist reviewed `d4a60691ee5f5149b91379e5f449e79828c5ed41` and left three comments:

- duplicate usage-limit regex in Claude should be extracted and synchronized
- duplicate usage-limit regex in Gemini should be extracted and synchronized
- `claude-result-handling` docs should mention `billing_mutation: "not_supported"`

Those comments were addressed in `466ad74bbdd0620ee60d82f695b9ca0d95871a8b`, replied to with commit/test evidence, and resolved. The follow-up Sonar duplication failure from the synchronized packaging copies was addressed in `635240497ba1a19b5245cbf07dbe98dcdb481a07`.

Internal current-head adversarial review focus:

- auth/session/provider-unavailable and transient tunnel/server failures must outrank quota-looking body text where status/context is available
- direct API generic 429 must remain `rate_limited`; direct API quota-bearing 429 must remain `usage_limited`
- direct API HTTP 501 must remain `provider_error`
- shared usage-limit detection must stay synchronized across packaged provider paths
- direct API and Grok `usage_limited` failures must produce `error_cause=cost_quota_usage_limit`
- runtime cost/quota diagnostics must be bounded to safe status/code/type metadata plus `billing_mutation`
- doctor/setup paths must not query billing endpoints or mutate account state
- PR #89 `git_binary_rejected` diagnostics must remain intact after conflict resolution

## Safety Notes

This PR does not add billing mutations, purchases, tier upgrades, payment flows, or account changes. It only classifies and reports provider cost/quota conditions using bounded metadata and explicit non-mutating readiness fields.

### CODE REVIEW

**Comment by seungpyoson** [comment]
Addressed the Greptile flat-response diagnostics mismatch and the SonarCloud reliability finding at latest head ad593a0.\n\nEvidence:\n- Direct API classification and cost_quota diagnostics now share the same provider failure detail extraction path, so flat bodies like { message: "quota exceeded" } keep error_code=usage_limited aligned with runtime_diagnostics.cost_quota.classification=usage_limited.\n- Added smoke coverage: direct API flat quota response keeps diagnostics aligned with error code.\n- Removed the duplicate Grok configured_timeout_ms diagnostic property that SonarCloud flagged as javascript:S1534.\n- Local verification: node --check for api-reviewer.mjs and grok-web-reviewer.mjs passed; node --test tests/smoke/api-reviewers.smoke.test.mjs passed 83/83; node --test tests/smoke/grok-web.smoke.test.mjs passed 61/61; pre-commit npm test passed 1013 total, 1007 pass, 0 fail, 6 skipped.\n\nGitHub checks for ad593a0 are queued/running now.

**Comment by seungpyoson** [comment]
Follow-up from latest-head DeepSeek adversarial review: verified and fixed the Grok/direct-API classification inconsistency.\n\nDeepSeek found that direct API detected quota/billing terms in response bodies, while Grok only classified usage_limited for HTTP 402/429. That was real: a tunnel-translated HTTP 400 with a quota body would have been tunnel_error/not_reported.\n\nFix at f69f803:\n- Grok now uses the same safe provider-detail extraction pattern for classification and cost_quota diagnostics.\n- Grok classifies HTTP 400 or other non-auth statuses with quota/billing body terms as usage_limited while keeping 401/403 as session_expired.\n- Added smoke coverage: custom-review maps HTTP 400 to usage_limited without leaking secrets.\n\nVerification:\n- node --check plugins/grok/scripts/grok-web-reviewer.mjs passed.\n- node --test tests/smoke/grok-web.smoke.test.mjs passed 62/62.\n- pre-commit npm test passed 1014 total, 1008 pass, 0 fail, 6 skipped.\n\nPrior external review evidence on ad593a0 is now stale for this small Grok follow-up; I will rerun latest-head review after GitHub checks settle.

**Comment by seungpyoson** [comment]
Latest-head external review update for f69f803 against base 8a0309b3:\n\nInitial full reviewer set on ad593a0:\n- DeepSeek completed: no blockers, but found the real Grok/direct-API body-classification inconsistency.\n- Gemini completed but self-caveated as a failed exact-diff slot due git execution permission block.\n- GLM failed with provider_unavailable/fetch failed after source selection.\n- Claude timed out at 600s after source transmission.\n- Kimi timed out at 600s after source transmission.\n\nFix landed at f69f803 for DeepSeek's concern, then targeted latest-head DeepSeek re-review completed with no blockers. DeepSeek verified Grok now classifies quota/billing response bodies consistently with direct API and keeps runtime_diagnostics.cost_quota aligned, including HTTP 400 quota body handling.\n\nLatest verification on f69f803:\n- Local pre-commit npm test: 1014 total, 1008 pass, 0 fail, 6 skipped.\n- Focused Grok smoke: 62/62.\n- GitHub checks: lint, test, smoke (api-reviewers/claude/gemini/grok/kimi), and SonarCloud all pass.\n\nRemaining reviewer notes are non-blocking follow-ups: possible generic PII redaction for provider error_message, finer Grok 4xx non-quota classification, and test-only requestFieldMatches hardening.

**Comment by seungpyoson** [comment]
Latest-head evidence for PR #90 after addressing the remaining Greptile P1:

- Current head is `90dd1f54da78b584ac69ce97599f229b8d1e6639` against base `8a0309b3b127d4f50f266ee61010eadcb1f5fb1b`.
- Fixed Grok classification ordering so `401`/`403` stay `session_expired` and `5xx` stays `tunnel_error` even when the response body contains billing/quota text.
- Fixed Grok `costQuotaDiagnostics` to report `classification: "not_reported"` for auth/server failures even with billing/quota body text.
- Added smoke coverage for `401` + quota body and `500` + quota body, while preserving `400`/`402`/`429` quota-body -> `usage_limited`.
- Local verification:
  - `node --check plugins/grok/scripts/grok-web-reviewer.mjs` passed.
  - `node --test tests/smoke/grok-web.smoke.test.mjs` passed: 62/62.
  - pre-commit `npm test` passed: 1014 total, 1008 pass, 0 fail, 6 skipped.
- GitHub CI passed on latest head: SonarCloud, lint, test, smoke api-reviewers, smoke claude, smoke gemini, smoke grok, smoke kimi.
- DeepSeek latest-head adversarial re-review completed with selected source sent; verdict: no blockers, coverage adequate, safe to merge from that review slot.

I am resolving the Greptile P1 thread tied to the 5xx/quota-body classification issue.

**Comment by seungpyoson** [comment]
Final review evidence for PR #90 at exact head `e85eca4f762d8938e711d1f7b09b725fc4ee3dd5` against base `8a0309b3b127d4f50f266ee61010eadcb1f5fb1b`.

Local verification:

- Focused API reviewer smoke: `84 passed, 0 failed`.
- Focused Grok smoke: `63 passed, 0 failed`.
- Full pre-commit suite: `1010 passed, 6 skipped, 0 failed`.
- GitHub checks are green: SonarCloud, lint, test, smoke api-reviewers, smoke claude, smoke gemini, smoke grok, smoke kimi.
- PR is mergeable.

External review matrix on the final head:

- GLM: completed, no blocking findings. Non-blocking follow-ups only.
- DeepSeek: completed, no blocking findings. Non-blocking follow-ups only.
- Gemini: completed with no code blockers or non-blocking concerns, but it reported a local environment/path limitation for direct git ref verification, so I am treating it as code-positive but not a perfect provenance slot.
- Kimi: completed and approved; no blockers.
- Claude: failed before inference under the required OAuth-only path. This is not a model verdict and not an approval. Hard evidence: OAuth status reports Claude Max login, but OAuth-only `claude -p` returns HTTP 401 with zero usage. Follow-up filed as #93.

Important Claude note:

I briefly proved that API-key auth can make Claude ping succeed, but that path violates the OAuth-only review policy and was not counted. The accidentally launched API-key diagnostic review was terminated and is not part of this evidence.

Greptile/Gemini/GLM follow-up handling:

- Direct API auth/session failures outrank quota-looking bodies.
- Direct API provider-unavailable statuses and unavailable/capacity body wording outrank quota-looking text in both `classifyHttpFailure` and `costQuotaDiagnostics`.
- Direct API includes HTTP 501 in provider-unavailable handling.
- Grok treats 408, 409, 425, and all 5xx as tunnel errors before quota-body matching.
- PII-shaped provider error tokens are dropped.
- Transient rate/capacity failure is not classified as cost-quota `usage_limited`.
- Doctor/setup paths do not probe or mutate billing.
- Billing/credit/tier changes remain explicit user-approved actions only.

Verdict: PR #90 is ready to present for merge consideration, with Claude OAuth-only readiness tracked separately in #93. I am not merging without explicit approval.

**Comment by seungpyoson** [comment]
@greptileai please re-review the current PR head only.

/gemini review

Exact refs for this re-review:
- Base: c6cebc4823d0a71b4798b07f7063215868f28709
- Head: be82eb2bf3a8929f2a3db34d161759dfb25b1006

Focus areas:
1. Cost/quota classification ordering: auth/session failures, transport errors, provider unavailable, and transient tunnel/server failures must not be misclassified as usage_limited just because the body contains quota-like text.
2. Direct API and Grok diagnostics: usage_limited should produce error_cause=cost_quota_usage_limit and safe runtime_diagnostics.cost_quota fields without leaking API keys, bearer tokens, account emails, plan IDs, payment IDs, or raw billing artifacts.
3. Billing safety: setup/doctor/readiness diagnostics must not call billing endpoints or attempt upgrades, purchases, or billing mutations.
4. Regression from PR #89 rebase: git_binary_rejected -> git_binary_policy must remain preserved in API Reviewers and Grok alongside usage_limited -> cost_quota_usage_limit.
5. Tests: verify behavior coverage, not only source-pattern checks, for API reviewer and Grok quota/provider-unavailable ordering.

Return blocking findings first with exact file/line evidence and repro path, then non-blocking concerns and test gaps.

File: plugins/api-reviewers/scripts/api-reviewer.mjs [hunk 1]
```diff
@@ -1371,16 +1383,59 @@ function providerErrorMessage(parsed, text, redact) {
...
+  return detail && typeof detail === "object" && !Array.isArray(detail) ? detail : {};
+}
+
 function classifyHttpFailure(status, parsed) {
-  const detail = parsed.ok ? JSON.stringify(parsed.value?.error ?? parsed.value ?? {}) : "";
+  const detail = parsed.ok ? providerFailureDetailText(parsed) : "";
   if (status === 401 || status === 403) return "auth_rejected";
-  if (status === 429) return "rate_limited";
-  if (status === 408 || status === 409 || status === 425 || status === 500 || status === 502 || status === 503 || status === 504 || /capacity|resource|overload|unavailable/i.test(detail)) {
+  if (status === 402 || status === 429) return "usage_limited";
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/api-reviewers/scripts/api-reviewer.mjs]: Fixed in d4a6069. Generic direct API HTTP 429 responses now classify as rate_limited, while 429 responses with quota/billing detail still classify as usage_limited. Added behavior coverage in tests/smoke/api-reviewers.smoke.test.mjs: direct API generic 429 rate limits stay rate_limited and direct API 429 quota status remains usage_limited. Verification: npm test passed, 1111 pass / 0 fail / 6 skipped.

File: plugins/api-reviewers/scripts/api-reviewer.mjs [hunk 2]
```diff
@@ -1371,16 +1383,59 @@ function providerErrorMessage(parsed, text, redact) {
...
+  if (status === 408 || status === 409 || status === 425 || status === 500 || status === 501 || status === 502 || status === 503 || status === 504 || /capacity|resource|overload|unavailable/i.test(detail)) {
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/api-reviewers/scripts/api-reviewer.mjs]: Fixed in d4a6069. HTTP 501 is no longer included in provider_unavailable classification; it now falls through to provider_error. Added smoke coverage: direct API 501 compatibility errors stay provider_error. Verification: npm test passed, 1111 pass / 0 fail / 6 skipped.

File: plugins/api-reviewers/scripts/api-reviewer.mjs [hunk 3]
```diff
@@ -1371,16 +1383,59 @@ function providerErrorMessage(parsed, text, redact) {
...
+  if (typeof value !== "string" && typeof value !== "number") return null;
+  const text = String(value);
+  return /^[A-Za-z0-9][A-Za-z0-9_.-]{0,63}$/.test(text) ? text : null;
+}
+
+function costQuotaDiagnostics(httpStatus, parsed) {
+  const error = providerFailureDetailObject(parsed);
+  const detail = parsed.ok ? providerFailureDetailText(parsed) : "";
+  const authRejected = httpStatus === 401 || httpStatus === 403;
+  const usageLimitStatus = httpStatus === 402 || httpStatus === 429;
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/api-reviewers/scripts/api-reviewer.mjs]: Fixed in d4a6069. costQuotaDiagnostics now matches classification: HTTP 429 is usage_limited only when the response body has quota/billing detail; generic 429 is cost_quota.classification=not_reported. Verification: npm test passed, 1111 pass / 0 fail / 6 skipped.

File: plugins/api-reviewers/scripts/api-reviewer.mjs [hunk 4]
```diff
@@ -1371,16 +1383,59 @@ function providerErrorMessage(parsed, text, redact) {
...
+  const providerUnavailable = !usageLimitStatus && (httpStatus === 408 || httpStatus === 409 || httpStatus === 425 || httpStatus === 500 || httpStatus === 501 || httpStatus === 502 || httpStatus === 503 || httpStatus === 504 || /capacity|resource|overload|unavailable/i.test(detail));
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/api-reviewers/scripts/api-reviewer.mjs]: Fixed in d4a6069. costQuotaDiagnostics no longer treats HTTP 501 as providerUnavailable; 501 now reports cost_quota.classification=not_reported and error_code=provider_error. Added smoke coverage for this exact case. Verification: npm test passed, 1111 pass / 0 fail / 6 skipped.

**Comment by seungpyoson** [comment]
@greptileai please re-review the current PR head only.

/gemini review

Exact refs for this re-review:
- Base: c6cebc4823d0a71b4798b07f7063215868f28709
- Head: d4a60691ee5f5149b91379e5f449e79828c5ed41

Delta since the prior Gemini Code Assist review:
- Generic direct API HTTP 429 now stays rate_limited.
- HTTP 429 with quota/billing detail remains usage_limited.
- HTTP 501 now falls through to provider_error, not provider_unavailable.
- runtime_diagnostics.cost_quota.classification now matches those decisions.

Focus areas:
1. Direct API classification ordering for 429, 501, auth failures, provider unavailable statuses, and quota-looking body text.
2. Safe runtime_diagnostics.cost_quota fields: no API keys, bearer tokens, account emails, plan IDs, payment IDs, or raw billing artifacts.
3. Regression from PR #89 rebase: git_binary_rejected -> git_binary_policy remains preserved alongside usage_limited -> cost_quota_usage_limit.
4. Test coverage: verify behavior tests cover generic 429, quota-bearing 429, and 501 compatibility errors.

Return blocking findings first with exact file/line evidence and repro path, then non-blocking concerns and test gaps.

File: plugins/claude/scripts/lib/claude.mjs [hunk 5]
```diff
@@ -136,19 +136,33 @@ export function parseClaudeResult(stdout) {
...
     raw: parsed,
   };
 }
 
+function usageLimitMessage(...values) {
+  const text = values.filter((value) => value != null).map((value) => (
+    typeof value === "string" ? value : JSON.stringify(value)
+  )).join("\n").trim();
+  if (!text) return null;
+  if (/(?:usage limit|quota|billing cycle|billing account|credit limit|insufficient credits)/i.test(text)) {
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/claude/scripts/lib/claude.mjs]: Fixed in 466ad74. Extracted the duplicated usage-limit detection into shared usage-limit.mjs copies, and Claude now imports usageLimitMessage from ./usage-limit.mjs. The shared helper includes payment_required, insufficient_quota, word-boundary quota/billing markers, credit/insufficient-credit markers, and Error code: 403 while excluding transient rate/capacity wording. Added behavior coverage in tests/unit/usage-limit.test.mjs and sync coverage in tests/unit/plugin-copies-in-sync.test.mjs. Verification: node --test focused suite passed; API/Grok smoke passed; npm test passed with 1126 pass / 0 fail / 6 skipped; pre-commit npm test passed.

File: plugins/claude/skills/claude-result-handling/SKILL.md [hunk 6]
```diff
@@ -226,9 +226,15 @@ from short-lived index contention.
...
-  credit-limit failure before returning JSON. Selected source content may have
-  been sent; render `error_summary`, `error_cause`, and `suggested_action`.
+- `usage_limited` — the target provider reported a quota, usage-tier,
+  billing-cycle, rate-limit, or credit-limit failure before returning a usable
+  review result. This can come from Claude, Gemini, Kimi, Grok, or the direct
+  API reviewers. Selected source content may have been sent; render
+  `error_summary`, `error_cause`, and `suggested_action`. Do not infer that the
+  plugin can buy credits, upgrade tiers, or change billing state. When present,
+  `runtime_diagnostics.cost_quota` contains safe provider status/code/type
+  metadata and `billing_mutation: "not_attempted"`; any financial
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/claude/skills/claude-result-handling/SKILL.md]: Fixed in 466ad74. Updated plugins/claude/skills/claude-result-handling/SKILL.md to state that runtime_diagnostics.cost_quota normally reports billing_mutation: not_attempted, and setup/doctor readiness can report billing_mutation: not_supported. Verification: tests/unit/docs-contracts.test.mjs passed in the focused suite, and npm test passed with 1126 pass / 0 fail / 6 skipped; pre-commit npm test passed.

File: plugins/gemini/scripts/lib/gemini.mjs [hunk 7]
```diff
@@ -58,41 +58,78 @@ function summarizeStderr(stderr) {
   return trimmed.length > 4000 ? `${trimmed.slice(0, 4000)}...` : trimmed;
 }
 
+function usageLimitMessage(...values) {
+  const text = values.filter((value) => value != null).map((value) => (
+    typeof value === "string" ? value : JSON.stringify(value)
+  )).join("\n").trim();
+  if (!text) return null;
+  if (/(?:usage limit|quota|billing cycle|billing account|credit limit|insufficient credits)/i.test(text)) {
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/gemini/scripts/lib/gemini.mjs]: Fixed in 466ad74. Extracted the duplicated usage-limit detection into shared usage-limit.mjs copies, and Gemini now imports usageLimitMessage from ./usage-limit.mjs. The shared helper includes payment_required, insufficient_quota, word-boundary quota/billing markers, credit/insufficient-credit markers, and Error code: 403 while excluding transient rate/capacity wording. Added behavior coverage in tests/unit/usage-limit.test.mjs and sync coverage in tests/unit/plugin-copies-in-sync.test.mjs. Verification: node --test focused suite passed; API/Grok smoke passed; npm test passed with 1126 pass / 0 fail / 6 skipped; pre-commit npm test passed.

**Comment by seungpyoson** [comment]
Addressed Greptile's remaining regex precision note in 05384c664e7e47bc174a2d3f7ca519402c34f0f9.\n\nTDD evidence:\n- Added regression coverage for `Error code: 4030` and `Error code: 40300` as non-usage-limit details while preserving plain `Error code: 403` as positive.\n- Confirmed RED first: `node --test tests/unit/usage-limit.test.mjs` failed because `Error code: 4030` returned true.\n- Fixed all synchronized `usage-limit.mjs` copies by changing `Error code:\s*403` to `Error code:\s*403\b`.\n\nVerification:\n- `node --test tests/unit/usage-limit.test.mjs` passed.\n- `node --test tests/unit/plugin-copies-in-sync.test.mjs` passed.\n- `node --test tests/smoke/api-reviewers.smoke.test.mjs` passed.\n- `node --test tests/smoke/grok-web.smoke.test.mjs` passed.\n- Pre-commit `npm test` passed: 1128 pass / 0 fail / 6 skipped.\n- GitHub checks pass on 05384c6: SonarCloud, lint, test, smoke api-reviewers, smoke claude, smoke gemini, smoke grok, smoke kimi.

File: plugins/api-reviewers/scripts/api-reviewer.mjs [hunk 8]
```diff
@@ -1363,24 +1433,62 @@ function parseJson(text) {
...
+  const detail = parsed.ok ? providerFailureDetailText(parsed) : String(text ?? "");
+  const usageLimitDetail = isUsageLimitDetail(detail);
+  if (status === 401 || (status === 403 && !usageLimitDetail)) return "auth_rejected";
+  if (status === 402 || (status === 403 && usageLimitDetail) || (status === 429 && usageLimitDetail)) return "usage_limited";
   if (status === 429) return "rate_limited";
   if (status === 408 || status === 409 || status === 425 || status === 500 || status === 502 || status === 503 || status === 504 || /capacity|resource|overload|unavailable/i.test(detail)) {
     return "provider_unavailable";
   }
+  if (isUsageLimitDetail(detail)) return "usage_limited";
   return "provider_error";
```
[the remark this answers is not in the document]
Comment (seungpyoson) [reply plugins/api-reviewers/scripts/api-reviewer.mjs:1476]: Fixed in d8d0b2c. Verified the finding was real at c99435a: HTTP 501 with quota text returned usage_limited. Added RED coverage by changing the 501 smoke payload to include insufficient_quota/quota text, then fixed classifyHttpFailure so 501 and unlisted 5xx responses cannot fall through to usage_limited. Added a 505 regression test for the wider unlisted-5xx case. Verification: node --test tests/smoke/api-reviewers.smoke.test.mjs passed 101/101; npm test passed 1155 pass / 0 fail / 6 skipped; pre-commit npm test passed.

**Comment by seungpyoson** [comment]
Latest-head internal adversarial review update for `535e957eda0b2855aa8c7242f16afd290a0a5ef2`:

- Internal review found one additional edge case after `d8d0b2c`: unlisted direct-API 5xx statuses could still be pulled into `provider_unavailable` when the body contained `unavailable` wording.
- TDD proof: changed `direct API unlisted 5xx quota-looking errors stay provider_error` to use HTTP 505 plus `quota endpoint unavailable...`; RED failed with `actual: provider_unavailable`, `expected: provider_error`.
- Fix: `classifyHttpFailure` now checks explicit transient statuses (`408`, `409`, `425`, `500`, `502`, `503`, `504`) as `provider_unavailable`, then pins any remaining `5xx` to `provider_error` before applying non-5xx unavailable/capacity wording.
- Verification:
  - Focused regression set passed: `501`, unlisted `5xx`, non-5xx provider-unavailable wording, and billing-provider outage tests.
  - `node --test tests/smoke/api-reviewers.smoke.test.mjs` passed: 101/101.
  - `CODEX_PLUGIN_SKIP_SMOKE=1 npm test` passed: 741 pass / 0 fail / 12 skipped.
  - GitHub checks on `535e957`: lint, test, smoke api-reviewers, smoke claude, smoke gemini, smoke grok, smoke kimi, and SonarCloud all pass.

Remaining gate: Greptile is still pending on latest head as of this comment.
